rotkehlchen/chain/bitcoin/utils.py

```diff
@@ -21,18 +30,90 @@ class OpCodes:
     op_checksig = b'\xac'
 
 
+class BitcoinSegwitEncoding(Enum):
+    """Enumeration type to list the various Bitcoin Segwit supported encodings."""
+    BECH32 = 1  # Native Segwit(P2WPKH)
+    BECH32M = 2  # Taproot(P2TR)
+
+
 def is_valid_btc_address(value: str) -> bool:
-    return is_valid_base58_address(value) or is_valid_bech32_address(value)
+    return (
+        is_valid_base58_address(value) or
+        is_valid_bech32_address(value) or
+        is_valid_bech32_bech32m_address(value)
+    )
 
 
 def is_valid_bech32_address(value: str) -> bool:
-    """Validates a bitcoin SegWit address for the mainnet
+    """Validates a bitcoin Segwit address using BIP-173
+    https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki
     """
-
     decoded = bech32.decode('bc', value)
     return decoded != (None, None)
 
 
+def is_valid_bech32_bech32m_address(value: str) -> bool:
+    """Validates a Bitcoin Taproot(P2TR) & Segwit(P2WPKH) address for the mainnet.
+
+    This validation is based on BIP-350 which improves on a flaw in BIP-173
+    https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki
+    """
+    hrpgot, data, spec = _bech32_decode(value)
+    if hrpgot != 'bc':
+        return False
+    if data is None:
+        return False
+    decoded = bech32.convertbits(data[1:], 5, 8, False)
+    if decoded is None or len(decoded) < 2 or len(decoded) > 40:
+        return False
+    if data[0] > 16:
+        return False
+    if data[0] == 0 and len(decoded) != 20 and len(decoded) != 32:
+        return False
+    if data[0] == 0 and spec != BitcoinSegwitEncoding.BECH32 or data[0] != 0 and spec != BitcoinSegwitEncoding.BECH32M:  # noqa: E501
+        return False
+    return (data[0], decoded) != (None, None)
+
+
+def _bech32_decode(value: str) -> Union[Tuple[str, List[int], BitcoinSegwitEncoding], Tuple[None, None, None]]:  # noqa: E501
```

Comment:
I think I explained that here https://github.com/rotki/rotki/pull/4714#discussion_r950178503

Reply:
But can you please add some part of that explanation in this docstring because not only Alexey, also I had the same question. And whoever else ends up reading this.

Reply:
Okay.